src/test/lib/utils/segment_tree_test.cpp

```diff
@@ -0,0 +1,38 @@
+#include "base_test.hpp"
+
+#include "utils/segment_tree.hpp"
+
+namespace hyrise {
+
+class SegmentTreeTest : public BaseTest {
+ public:
+  static auto leaf_count(const auto& segment_tree) {
+    return segment_tree.leaf_count;
+  }
+
+  static auto tree_size(const auto& segment_tree) {
+    return segment_tree.tree.size();
+  }
+
+  SegmentTree<int, std::plus<>> segment_tree{std::array{1, 2, 3}, std::plus<>{}, element_factory(0)};
+};
+
+TEST_F(SegmentTreeTest, BinaryTreeIsFull) {
+  EXPECT_EQ(leaf_count(segment_tree), 4);
+  // 8 nodes, as index 0 is unused
+  EXPECT_EQ(tree_size(segment_tree), 9);
```

Comment:
See the comment in the `SegmentTree` class. We've got the feeling that one node too many gets allocated.